### DISCUSSION TITLE
register multiple data sources in jdo config

### DISCUSSION THREAD

**Discussion by nyemiks** [openmdx/openmdx#49] [Q&A] [answered]
I have an app that is going to have to access two different opencrx databases. 
In my app server, I can create a data source for each database then map them to a jndi that I define for each of them.

is there a configuration option that allows me to register all my data sources (with their jndi name) in the jdo config ? or any other configuration file in openmdx.

**Comment by nyemiks** [comment]
Would really appreciate help from the community on this issue. I am stuck.

**Comment by brunoamacos** [comment]
I'm a bit confused by your question. Do you mean

- A 3rd-Party app that accesses to different databases without opencrx api, only sql? Then it's a jdbc problem, you need to configure to different datasources and access them with appropriate jndi lookups

- A 3rd-Party app that accesses to different opencrx instances through the rest api and the opencrx api. Than you should have a look at opencrx-client examples. You can have multiple Persistence Managers like this with different urls, principals etc.
```java
PersistenceManagerFactory pmf = org.opencrx.kernel.utils.Utils.getPersistenceManagerFactoryProxy(REST_URL,
        PRINCIPAL_NAME,
        PASSWORD,
        "application/vnd.openmdx.wbxml");
PersistenceManager pm = pmf.getPersistenceManager(PRINCIPAL_NAME,
        PASSWORD);
```
- two different opencrx instances and a 3rd Party app, all in the same application (ear). I dont think that this whould work.

**Reply by nyemiks** [reply]
None of the above 

my scenario is as follows:

My app is opencrx based. (i.e. Uses the opencrx api natively. I am not using the rest client api.)  This apps needs to access two different opencrx databases.  This app runs on an application server (Glassfish, payara, tomcat e.t.c).
Since it runs in an app server I have two create data sources to connect to the databases. I also configure jndi.

My question is How can I obtain a persistence manager factory for each database ?

To provide further context

For another app that I developed. The app uses only one opencrx database so I get my persistence manager factory as follows

PersistenceManagerFactory pmf = Utils.getPersistenceManagerFactory();

The above code works provided that I have created a data source with the jndi name "jdbc_opencrx_CRX". (I think this name is hard
coded somewhere in the code because it does not work if I put another name. )

Now I need my app to access two different opencrx databases. How can I get a persistence manager factory for each database ?

I tried the below code 

PersistenceManagerFactory pmf = (PersistenceManagerFactory) new InitialContext().lookup(jndiName); 

It does not work.

**Comment by wfro64** [comment, marked as answer]
The configuration option is `PERSISTENCE/datasourceName[0]`. It is located in:

* `APP-INF/lib/opencrx-config-crx.jar/META-INF/Kernel.properties`
* `APP-INF/lib/opencrx-config-crx.jar/META-INF/Security.properties`

These property files are referenced in `APP-INF/lib/opencrx-config-crx.jar/META-INF/jdoconfig.xml`:

```
<!--  Dataprovider Kernel -->             
<property name="org.opencrx.kernel.rest.PlugIn.Kernel.class" value="org.openmdx.base.dataprovider.kernel.Dataprovider_2"/>
<property name="org.opencrx.kernel.rest.PlugIn.Kernel.configuration" value="xri://+resource/META-INF/Kernel.properties"/>

<!--  Dataprovider Security -->             
<property name="org.opencrx.kernel.rest.PlugIn.Security.class" value="org.openmdx.base.dataprovider.kernel.Dataprovider_2"/>
<property name="org.opencrx.kernel.rest.PlugIn.Security.configuration" value="xri://+resource/META-INF/Security
```

You have to deploy an openCRX EAR for each configured datasource.
